docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+     types.BlockID, height int64, commit *types.Commit) error` in the
+     current Golang implementation, which expects blockID and height 
+	 (from the first block) and the
+     corresponding commit from the following block. We use the
+     simplified form for ease in presentation.
+
+- Implementation remark
+    <!-- - implements the check from -->
+    <!--  [**[TMBC-SOUND-DISTR-PossCommit]**][TMBC-SOUND-DISTR-PossCommit--link], -->
+    <!--  that is, that  *c* is a valid commit for block *b* -->
+	- implements the check that  *c* is a valid commit for block *b*
+- Expected precondition
+    - *c* is a valid commit for block *b*
+- Expected postcondition
+    - *true* if precondition holds
+	- *false* if precondition is violated
+- Error condition
+    - none
+----
+
+## Algorithm Invariants
+
+> In contrast to the temporal properties above that define the problem
+> statement, the following are invariants on the solution to the
+> problem, that is on the algorithm. These invariants are useful for
+> the verification, but can also guide the implementation.
+
+#### **[FS-VAR-STATE-INV]**:
+It is always the case that *state* corresponds to the application state of the
+blockchain of that height, that is, *state = chain[height -
+1].AppState*; *chain* is defined in
+[**[TMBC-SEQ]**][TMBC-SEQ-link].
+
+#### **[FS-VAR-PEER-INV]**:
+It is always the case that the set *peerIDs* only contains nodes that
+have not yet misbehaved (by sending wrong data or timing out).
+
+#### **[FS-VAR-BLOCK-INV]**:
+For *startBlock.Height <= i < height - 1*, let *b(i)* be the block with
+height *i* in *blockstore*, it always holds that
+*VerifyCommit(b(i), b(i+1).Commit) = true*. This means that *height*
+can only be incremented if all blocks with lower height have been verified.
+
+
+## FastSync V2
+
+
+<!--
+> Basic data structures. Simplified, so that we can focus on the distributed
+algorithm here. If existing: link to Tendermint data structures, and mentioned
+if details were omitted. 
+---->
+
+### Outline
+
+<!--
+> Describe solution (in English), decomposition into functions, where communication to other components happens.
+---->
+
+The protocol is described in terms of functions that are triggered by
+(external) events. The implementation uses a scheduler and a
+de-multiplexer to deal with communicating with peers and to
+trigger the execution of these functions:
+
+- `QueryStatus()`: regularly (currently every 10sec; necessarily
+  interval greater than *2 Delta*) queries all peers from *peerIDs*
+  for their current height [TMBC-CORR-FULL]. It does so
+  by calling `Status(n)` remotely on all peers *n*.
+  
+- `CreateRequest`: regularly checks whether certain blocks have no
+  open request. If a block does not have an open request, it requests
+  one from a peer. It does so by calling `Block(n,h)` remotely on one
+  peer *n* for a missing height *h*.
+  
+> We have left the strategy how peers are selected unspecified, and
```

Comment:
Maybe we should explain the strategy used as it is part of the solution. It basically selects a peer with the minimum number of pending requests that can serve given height h (peerHeight >= h).

Reply:
I added the clarification for V2 to the comment.